File: test-platform-v2/backend/app/api/v2/production.py
"""AITDE V3.6 production API (V36-013/014).

Production is an Evidence Source, not a Test Runtime. All endpoints operate in a
read-only posture enforced by the policy/service layer; the security boundary is
``ReadOnlyBrowserPolicy`` / ``ProductionDbGuard`` + a truly read-only DB account,
never the UI. Sensitive values are never returned.
"""
from __future__ import annotations

import json

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.v2.deps import require_aitde_v3
from app.core.deps import CurrentUser, get_current_user, get_db
from app.modules.aitde.production import repository as repo
from app.modules.aitde.production import services, schemas
from app.schemas.common import R

router = APIRouter(
    prefix="/production",
    tags=["AITDE - Production Evidence"],
    dependencies=[Depends(require_aitde_v3)],
)


@router.post("/observation-sessions", response_model=R[dict])
def start_observation_session(
    payload: schemas.ObservationSessionCreate,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    sid = services.production_observer_service.start(
        db,
        project_id=payload.project_id,
        environment_id=payload.environment_id,
        mission_id=payload.mission_id,
        worker_id=payload.worker_id,
        mode=payload.mode.value,
        started_by=payload.started_by or current.user.id,
        policy_version=payload.policy_version,
    )
    return R.ok({"id": sid})


@router.post("/observation-sessions/{session_id}/stop", response_model=R[dict])
def stop_observation_session(
    session_id: int,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    services.production_observer_service.stop(db, session_id, user_id=current.user.id)
    return R.ok({"id": session_id})


@router.get("/observation-sessions/{session_id}", response_model=R[dict])
def get_observation_session(
    session_id: int,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    return R.ok(services.production_observer_service.status(db, session_id))


@router.get("/journeys", response_model=R[list])
def list_journeys(
    session_id: int | None = None,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    rows = repo.list_journeys(db, current.project_id or 0, session_id)
    return R.ok(
        [
            {
                "id": r.id,
                "project_id": r.project_id,
                "session_id": r.session_id,
                "name": r.name,
                "journey_hash": r.journey_hash,
                "summary_json": r.summary_json,
                "source_ref_json": r.source_ref_json,
                "created_at": r.created_at.isoformat() if r.created_at else None,
            }
            for r in rows
        ]
    )


@router.get("/journeys/{journey_id}", response_model=R[dict])
def get_journey(
    journey_id: int,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    row = repo.get_journey(db, journey_id)
    if row is None:
        return R.err(code=404, msg="journey not found")
    steps = repo.list_journey_steps(db, journey_id)
    return R.ok(
        {
            "id": row.id,
            "project_id": row.project_id,
            "session_id": row.session_id,
            "name": row.name,
            "journey_hash": row.journey_hash,
            "summary_json": json.loads(row.summary_json or "{}"),
            "source_ref_json": json.loads(row.source_ref_json or "{}"),
            "created_at": row.created_at.isoformat() if row.created_at else None,
            "steps": [
                {
                    "sequence": s.sequence,
                    "event_type": s.event_type,
                    "semantic_action": json.loads(s.semantic_action_json or "{}"),
                    "url_template": s.url_template,
                    "xhr_refs": json.loads(s.xhr_refs_json or "{}"),
                    "evidence_refs": json.loads(s.evidence_refs_json or "[]"),
                }
                for s in steps
            ],
        }
    )


@router.post("/data/inspect", response_model=R[dict])
def inspect_data(
    payload: schemas.DbInspectRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    # The row provider is a no-op for a plain API call; service-layer guards and
    # query audit still run. Real probes are driven by a capability worker.
    result = services.production_db_explorer.inspect(
        db,
        project_id=payload.project_id,
        data_source_id=payload.data_source_id,
        session_id=payload.session_id,
        sql=payload.sql,
        schema=payload.schema_name,
        table_names=payload.table_names,
    )
    return R.ok(result)


@router.post("/entity-graphs/extract", response_model=R[dict])
def extract_entity_graph(
    payload: schemas.EntityGraphExtractRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    graph, content_hash = services.entity_graph_extractor.extract(
        root_entity_type=payload.root_entity_type,
        root_ref_hash=payload.root_ref_hash,
        child_loader=lambda *a: [],
    )
    row = repo.create_entity_graph_snapshot(
        db,
        {
            "project_id": payload.project_id,
            "mission_id": payload.mission_id,
            "source_environment_id": payload.source_environment_id,
            "root_entity_type": payload.root_entity_type,
            "root_ref_hash": payload.root_ref_hash,
            "graph_json": json.dumps(graph, ensure_ascii=False),
            "content_hash": content_hash,
        },
    )
    return R.ok({"id": row.id, "content_hash": content_hash, "nodes": len(graph["nodes"]), "edges": len(graph["edges"])})


@router.post("/templates", response_model=R[dict])
def build_template(
    payload: schemas.TemplateBuildRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    snapshot = repo.get_entity_graph_snapshot(db, payload.entity_graph_snapshot_id)
    if snapshot is None:
        return R.err(code=404, msg="entity graph snapshot not found")
    rules = repo.list_masking_rules(db, payload.masking_profile_id) if payload.masking_profile_id else []
    template = services.prod_template_builder.build(
        name=payload.name,
        graph=json.loads(snapshot.graph_json or "{}"),
        masking_profile_id=payload.masking_profile_id,
        rules=rules,
        project_id=payload.project_id,
        mission_id=payload.mission_id,
        created_by=payload.created_by or current.user.id,
    )
    row = repo.create_prod_template(
        db,
        {
            "project_id": payload.project_id,
            "mission_id": payload.mission_id,
            "name": payload.name,
            "entity_graph_snapshot_id": payload.entity_graph_snapshot_id,
            "masking_profile_id": payload.masking_profile_id,
            "template_json": json.dumps(template, ensure_ascii=False),
            "validation_status": "PENDING",
            "created_by": payload.created_by or current.user.id,
        },
    )
    return R.ok({"id": row.id, "validation_status": row.validation_status})


@router.post("/templates/{template_id}/validate", response_model=R[dict])
def validate_template(
    template_id: int,
    payload: schemas.TemplateValidateRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    row = repo.get_prod_template(db, template_id)
    if row is None:
        return R.err(code=404, msg="template not found")
    template = json.loads(row.template_json or "{}")
    report = services.masking_service.validation_report(
        profile_id=row.masking_profile_id or 0, record=template
    )
    status = "VALID" if report["valid"] else "INVALID"
    repo.update_prod_template(db, template_id, {"validation_status": status})
    return R.ok({"validation_status": status, "leaks": report["leaks"]})


@router.post("/templates/{template_id}/materialize", response_model=R[dict])
def materialize_template(
    template_id: int,
    payload: schemas.TemplateMaterializeRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    mid = services.template_materializer.materialize(
        db,
        template_id=template_id,
        target_environment_id=payload.target_environment_id,
        project_id=payload.project_id,
    )
    return R.ok({"materialization_id": mid})


@router.post("/evidence/{journey_id}/analyze-gaps", response_model=R[dict])
def analyze_gaps(
    journey_id: int,
    payload: schemas.GapAnalysisRequest,
    current: CurrentUser = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    row = repo.get_journey(db, journey_id)
    if row is None:
        return R.err(code=404, msg="journey not found")
    journey = {
        "events": [
            {
                "url": s.url_template,
                "semantic_action": json.loads(s.semantic_action_json or "{}"),
            }
            for s in repo.list_journey_steps(db, journey_id)
        ]
    }
    proposals = services.production_evidence_to_design_service.analyze_gaps(journey=journey)
    return R.ok(
        [
            {"kind": p["kind"], "title": p["title"], "confidence": p["confidence"], "auto_approved": p.get("auto_approved", False)}
            for p in proposals
        ]
    )

File: test-platform-v2/frontend/src/api/production.ts
import { aitdeV2 } from './missions'

// ── AITDE V3.6 Production Evidence & Real-World Data Template domain types ──

/** A JSON-ish value that may arrive already parsed or as a raw string. */
export type JsonLike = Record<string, unknown> | string | null

export interface ObservationSession {
  id: number
  project_id: number
  mission_id?: number | null
  environment_id: number
  worker_id?: number | null
  mode: string
  status: string
  policy_version?: string | null
  started_by?: number | null
  started_at?: string | null
  finished_at?: string | null
}

export interface StartObservationSessionInput {
  project_id: number
  environment_id: number
  mission_id?: number | null
  worker_id?: number | null
  mode: string
  started_by?: number | null
  policy_version?: string | null
}

export interface Journey {
  id: number
  project_id: number
  session_id: number
  name: string
  journey_hash: string
  summary_json: JsonLike
  source_ref_json: JsonLike
  created_at?: string | null
}

export interface XhrRef {
  method?: string
  status?: number | string | null
  url?: string | null
  request_url?: string | null
  request_headers?: JsonLike
  request_body?: unknown
  response_headers?: JsonLike
  response_body?: unknown
  headers?: JsonLike
  body?: unknown
  [key: string]: unknown
}

export interface JourneyStep {
  sequence: number
  event_type: string
  semantic_action?: Record<string, unknown> | string | null
  url_template?: string | null
  xhr_refs?: XhrRef[] | Record<string, unknown> | string | null
  evidence_refs?: unknown[] | string | null
}

export interface JourneyDetail extends Journey {
  steps: JourneyStep[]
}

export interface InspectDataInput {
  project_id: number
  data_source_id: number
  sql: string
  schema_name?: string | null
  session_id?: number | null
  table_names?: string[]
}

export interface InspectDataResult {
  rows: unknown[]
  row_count: number
  duration_ms: number
}

export interface ExtractEntityGraphInput {
  project_id: number
  root_entity_type: string
  root_ref_hash: string
  source_environment_id: number
  mission_id?: number | null
}

export interface EntityNode {
  id: string
  label?: string | null
  entity_type?: string | null
  ref_hash?: string | null
  [key: string]: unknown
}

export interface EntityEdge {
  source: string
  target: string
  relation?: string | null
  [key: string]: unknown
}

export interface EntityGraphResult {
  id: number
  content_hash: string
  nodes: EntityNode[]
  edges: EntityEdge[]
}

export interface BuildTemplateInput {
  project_id: number
  name: string
  entity_graph_snapshot_id: number
  masking_profile_id?: number | null
  mission_id?: number | null
  created_by?: number | null
}

export interface BuildTemplateResult {
  id: number
  validation_status: string
}

export interface ValidateTemplateInput {
  project_id: number
  template_id: number
}

export interface ValidateTemplateResult {
  validation_status: string
  leaks: unknown[]
}

export interface MaterializeTemplateInput {
  project_id: number
  template_id: number
  target_environment_id: number
}

export interface MaterializeTemplateResult {
  materialization_id: number
}

export interface AnalyzeGapsInput {
  project_id: number
  journey_id: number
}

export interface GapCandidate {
  kind: string
  title: string
  confidence: number
  auto_approved: boolean
}

/** A masking rule that composes a masking profile (client-side DSL). */
export interface MaskingRule {
  id: string
  entity_pattern: string
  field_pattern: string
  classification: string
  strategy: 'REDACT' | 'HASH' | 'TOKENIZE' | 'FAKE' | 'PRESERVE'
  priority: number
}

/** A masking profile being composed in the masking UI. */
export interface MaskingProfile {
  id: string
  name: string
  description?: string | null
  rules: MaskingRule[]
}

// ── Observation sessions ──

export function startObservationSession(
  payload: StartObservationSessionInput,
): Promise<{ id: number }> {
  return aitdeV2.post('/production/observation-sessions', payload)
}

export function stopObservationSession(id: number): Promise<{ id: number }> {
  return aitdeV2.post(`/production/observation-sessions/${id}/stop`)
}

export function fetchObservationSession(id: number, signal?: AbortSignal): Promise<ObservationSession> {
  return aitdeV2.get(`/production/observation-sessions/${id}`, { signal })
}

// ── Journeys ──

export function fetchJourneys(sessionId?: number, signal?: AbortSignal): Promise<Journey[]> {
  return aitdeV2.get('/production/journeys', {
    params: sessionId != null ? { session_id: sessionId } : undefined,
    signal,
  })
}

export function fetchJourney(id: number, signal?: AbortSignal): Promise<JourneyDetail> {
  return aitdeV2.get(`/production/journeys/${id}`, { signal })
}

// ── Real-world data ──

export function inspectProductionData(payload: InspectDataInput): Promise<InspectDataResult> {
  return aitdeV2.post('/production/data/inspect', payload)
}

// ── Entity graphs (extract) ──

export function extractEntityGraph(payload: ExtractEntityGraphInput): Promise<EntityGraphResult> {
  return aitdeV2.post('/production/entity-graphs/extract', payload)
}

// ── Templates (build / validate / materialize) ──

export function buildTemplate(payload: BuildTemplateInput): Promise<BuildTemplateResult> {
  return aitdeV2.post('/production/templates', payload)
}

export function validateTemplate(
  id: number,
  payload: ValidateTemplateInput,
): Promise<ValidateTemplateResult> {
  return aitdeV2.post(`/production/templates/${id}/validate`, payload)
}

export function materializeTemplate(
  id: number,
  payload: MaterializeTemplateInput,
): Promise<MaterializeTemplateResult> {
  return aitdeV2.post(`/production/templates/${id}/materialize`, payload)
}

// ── Evidence gap analysis ──

export function analyzeGaps(journeyId: number, payload: AnalyzeGapsInput): Promise<GapCandidate[]> {
  return aitdeV2.post(`/production/evidence/${journeyId}/analyze-gaps`, payload)
}

// ── Label maps ──

export const SESSION_STATUS_LABELS: Record<string, { label: string; color: string }> = {
  ACTIVE: { label: '采集中', color: 'bg-status-info-muted text-status-info' },
  COMPLETED: { label: '已完成', color: 'bg-status-success-muted text-status-success' },
  COMPLETE: { label: '已完成', color: 'bg-status-success-muted text-status-success' },
  STOPPED: { label: '已停止', color: 'bg-status-warning-muted text-status-warning' },
  FAILED: { label: '失败', color: 'bg-status-danger-muted text-status-danger' },
}

export const OBSERVATION_MODE_LABELS: Record<string, { label: string; color: string }> = {
  OBSERVE: { label: '观察', color: 'bg-status-info-muted text-status-info' },
  REAL: { label: '真实流量', color: 'bg-status-warning-muted text-status-warning' },
  REPLAY: { label: '回放', color: 'bg-status-info-muted text-status-info' },
  STAGING: { label: '预发布', color: 'bg-status-warning-muted text-status-warning' },
  LIVE: { label: '线上', color: 'bg-status-danger-muted text-status-danger' },
}

export const TEMPLATE_VALIDATION_LABELS: Record<string, { label: string; color: string }> = {
  VALID: { label: 'Valid', color: 'bg-status-success-muted text-status-success' },
  OK: { label: 'Valid', color: 'bg-status-success-muted text-status-success' },
  INVALID: { label: 'Invalid', color: 'bg-status-danger-muted text-status-danger' },
  LEAKS_FOUND: { label: '发现泄漏', color: 'bg-status-danger-muted text-status-danger' },
  VALIDATING: { label: '校验中', color: 'bg-status-warning-muted text-status-warning' },
  DRAFT: { label: '草稿', color: 'bg-muted text-muted-foreground' },
}
